### PR TITLE
RavenDB-14472 - Server wide external replication

### PR DESCRIPTION
**Pull request by grisha-kotler**
https://issues.hibernatingrhinos.com/issue/RavenDB-14472

### CODE REVIEW

**Review by ml054 — APPROVED** [review]
studio changes looks good

File: src/Raven.Client/Documents/Operations/Replication/ServerWideExternalReplication.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Raven.Client.Documents.Operations.Replication
+{
+    public class ServerWideExternalReplication : ExternalReplication, IDynamicJson
+    {
+        internal static string NamePrefix = "Server Wide External Replication";
+
+        internal static string RavenConnectionStringPrefix = "Server Wide Raven Connection String";
+
+        public string[] TopologyDiscoveryUrls;
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Replication/ServerWideExternalReplication.cs:11]: convert to property

File: src/Raven.Client/ServerWide/Operations/Configuration/DeleteServerWideTaskOperation.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Net.Http;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Operations.OngoingTasks;
+using Raven.Client.Http;
+using Raven.Client.Util;
+using Sparrow.Json;
+
+namespace Raven.Client.ServerWide.Operations.Configuration
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/DeleteServerWideTaskOperation.cs:9]: Raven.Client.Documents.Operations.OngoingTasks

File: src/Raven.Client/ServerWide/Operations/Configuration/DeleteServerWideTaskOperation.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+{
+    public class DeleteServerWideTaskOperation : IServerOperation
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/DeleteServerWideTaskOperation.cs:11]: DeleteServerWideOngoingTaskOperation

File: src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideBackupConfigurationOperation.cs [hunk 4]
```diff
@@ -34,7 +35,7 @@ public GetServerWideBackupConfigurationCommand(string name)
 
             public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
             {
-                url = $"{node.Url}/admin/configuration/server-wide/backup?name={Uri.EscapeDataString(_name)}";
+                url = $"{node.Url}/admin/configuration/server-wide/tasks?type={OngoingTaskType.Backup}&name={Uri.EscapeDataString(_name)}";
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideBackupConfigurationOperation.cs:38]: This breaks the backward compatibility

File: src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideBackupConfigurationsOperation.cs [hunk 5]
```diff
@@ -19,7 +20,7 @@ private class GetServerWideBackupConfigurationsCommand : RavenCommand<ServerWide
 
             public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, ServerNode node, out string url)
             {
-                url = $"{node.Url}/admin/configuration/server-wide/backup";
+                url = $"{node.Url}/admin/configuration/server-wide/tasks?type={OngoingTaskType.Backup}";
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideBackupConfigurationsOperation.cs:23]: This breaks the backward compatibility

File: src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideExternalReplicationOperation.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Net.Http;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Operations.OngoingTasks;
+using Raven.Client.Documents.Operations.Replication;
+using Raven.Client.Http;
+using Raven.Client.Json.Converters;
+using Sparrow.Json;
+
+namespace Raven.Client.ServerWide.Operations.Configuration
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideExternalReplicationOperation.cs:10]: .Replication?

File: src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideExternalReplicationsOperation.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
+﻿using System.Net.Http;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Operations.OngoingTasks;
+using Raven.Client.Documents.Operations.Replication;
+using Raven.Client.Http;
+using Raven.Client.Json.Converters;
+using Sparrow.Json;
+
+namespace Raven.Client.ServerWide.Operations.Configuration
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideExternalReplicationsOperation.cs:9]: .Replication?

File: src/Raven.Client/ServerWide/Operations/Configuration/PutServerWideBackupConfigurationOperation.cs [hunk 8]
```diff
@@ -63,12 +63,13 @@ public override void SetResponse(JsonOperationContext context, BlittableJsonRead
             }
         }
     }
-    
-    public class PutServerWideBackupConfigurationResponse : IDynamicJson 
+
+    public class ServerWideTaskResponse : IDynamicJson
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/PutServerWideBackupConfigurationOperation.cs:67]: ServerWideOngoingTaskResponse

Comment (ppekrol) [reply src/Raven.Client/ServerWide/Operations/Configuration/PutServerWideBackupConfigurationOperation.cs:67]: Move to separate file

File: src/Raven.Client/ServerWide/Operations/Configuration/PutServerWideExternalReplicationOperation.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Operations.Replication;
+using Raven.Client.Documents.Session;
+using Raven.Client.Http;
+using Raven.Client.Json;
+using Raven.Client.Json.Converters;
+using Raven.Client.Util;
+using Sparrow.Json;
+
+namespace Raven.Client.ServerWide.Operations.Configuration
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/PutServerWideExternalReplicationOperation.cs:12]: .Replication?

File: src/Raven.Client/ServerWide/Operations/Configuration/ToggleServerWideTaskStateOperation.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Net.Http;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Documents.Operations.OngoingTasks;
+using Raven.Client.Http;
+using Raven.Client.Util;
+using Sparrow.Json;
+
+namespace Raven.Client.ServerWide.Operations.Configuration
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/ToggleServerWideTaskStateOperation.cs:9]: .OngoingTasks

File: src/Raven.Client/ServerWide/Operations/Configuration/ToggleServerWideTaskStateOperation.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+{
+    public class ToggleServerWideTaskStateOperation : IServerOperation
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/ToggleServerWideTaskStateOperation.cs:11]: ToggleServerWideOngoingTaskStateOperation

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 12]
```diff
@@ -72,7 +73,25 @@ public class ClusterStateMachine : RachisStateMachine
         public static readonly TableSchema IdentitiesSchema;
         public static readonly TableSchema CertificatesSchema;
 
-        public static string ServerWideBackupConfigurationsKey = "server-wide/backup/configurations";
+        public class ServerWideConfigurationKey
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs:76]: static

File: src/Raven.Server/ServerWide/Commands/DeleteServerWideBackupConfigurationCommand.cs [hunk 13]
```diff
@@ -12,10 +13,10 @@ protected DeleteServerWideBackupConfigurationCommand()
             // for deserialization
         }
 
+        [Obsolete]
         public DeleteServerWideBackupConfigurationCommand(string configurationName, string uniqueRequestId) : base(uniqueRequestId)
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/DeleteServerWideBackupConfigurationCommand.cs:17]: You can remove that ctor I think?

File: src/Raven.Server/ServerWide/Commands/DeleteServerWideTaskCommand.cs [hunk 14]
```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using System.Collections.Generic;
+using Raven.Client.Documents.Operations.OngoingTasks;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.ServerWide.Commands
+{
+    public class DeleteServerWideTaskCommand : UpdateValueCommand<DeleteServerWideTaskCommand.DeleteConfiguration>
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/DeleteServerWideTaskCommand.cs:10]: DeleteServerWideOngoingTaskCommand

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs [hunk 15]
```diff
@@ -87,89 +89,150 @@ public async Task PutServerWideBackupConfigurationCommand()
...
+            await DeleteServerWideTaskCommand(OngoingTaskType.Backup);
+        }
+
+        [RavenAction("/admin/configuration/server-wide/task", "DELETE", AuthorizationStatus.ClusterAdmin)]
+        public async Task DeleteServerWideTaskCommand()
+        {
+            var typeAsString = GetStringQueryString("type", required: true);
+
+            if (Enum.TryParse(typeAsString, out OngoingTaskType type) == false)
+                throw new ArgumentException($"{typeAsString} is unknown task type.");
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminServerWideHandler.cs:138]: BadRequestException

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs [hunk 16]
```diff
@@ -87,89 +89,150 @@ public async Task PutServerWideBackupConfigurationCommand()
...
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException($"Task type '{type} isn't suppported");
+            }
+
+            GetTaskConfigurations(type, converter);
+            return Task.CompletedTask;
+        }
+
+        [RavenAction("/admin/configuration/server-wide/tasks-for-studio", "GET", AuthorizationStatus.ClusterAdmin)]
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminServerWideHandler.cs:175]: What is this? We have dedicated Handlers for Studio items.

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs [hunk 17]
```diff
@@ -87,89 +89,150 @@ public async Task PutServerWideBackupConfigurationCommand()
...
+        public async Task ToggleServerWideTaskState()
         {
+            var typeAsString = GetStringQueryString("type", required: true);
+            var taskName = GetStringQueryString("name", required: true);
             var disable = GetBoolValueQueryString("disable") ?? true;
-            var taskName = GetStringQueryString("taskName", required: false);
-           
+
+            if (Enum.TryParse(typeAsString, out OngoingTaskType type) == false)
+                throw new ArgumentException($"{typeAsString} is unknown task type.");
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminServerWideHandler.cs:219]: BadRequestException

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs [hunk 18]
```diff
@@ -191,17 +254,71 @@ public async Task FullBackupDataDirectory()
...
+
+                context.Write(writer, result.ToJson());
+                writer.Flush();
+            }
+        }
     }
-    
-    public class ServerWideBackupConfigurationResults : IDynamicJson
+
+    public class ServerWideTasksResult<T> : IDynamicJson
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminServerWideHandler.cs:312]: ServerWideOngoingTasksResult

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs [hunk 19]
```diff
@@ -210,11 +327,35 @@ public DynamicJsonValue ToJson()
             };
         }
     }
+
+    public class ServerWideTaskConfigurations : IDynamicJson
```
Comment (ppekrol) [inline src/Raven.Server/Web/System/AdminServerWideHandler.cs:331]: ServerWideOngoingTaskConfigurations

File: src/Raven.Server/ServerWide/Commands/PutServerWideExternalReplicationCommand.cs [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            return $"{ServerWideExternalReplication.RavenConnectionStringPrefix} for {serverWideName}";
+        }
+
+        public static RavenConnectionString UpdateExternalReplicationTemplateForDatabase(ExternalReplication configuration, string databaseName, string[] topologyDiscoveryUrls)
+        {
+            var serverWideName = configuration.Name;
+            configuration.Name = GetTaskName(serverWideName);
+            configuration.Database = databaseName;
+            configuration.ConnectionStringName = GetRavenConnectionStringName(serverWideName);
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/PutServerWideExternalReplicationCommand.cs:96]: Is this needed when 'topologyDiscoveryUrls' are passed externally?

File: src/Raven.Client/Documents/Operations/Replication/ServerWideExternalReplication.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Operations/Replication/ServerWideExternalReplication.cs:11]: The regular external replication handles that via ConnectionStringName, do you think we can do something similar?

**Review by ppekrol — COMMENTED** [review]
This requires tests

File: src/Raven.Client/Documents/Operations/Replication/ServerWideExternalReplication.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Client/Documents/Operations/Replication/ServerWideExternalReplication.cs:11]: done

File: src/Raven.Client/ServerWide/Operations/Configuration/DeleteServerWideTaskOperation.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Client/ServerWide/Operations/Configuration/DeleteServerWideTaskOperation.cs:9]: done

File: src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideBackupConfigurationOperation.cs (hunk 4 above)
Comment (grisha-kotler) [reply src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideBackupConfigurationOperation.cs:38]: I left the endpoint for backward compatibility

File: src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideBackupConfigurationsOperation.cs (hunk 5 above)
Comment (grisha-kotler) [reply src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideBackupConfigurationsOperation.cs:23]: Same here

File: src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideExternalReplicationOperation.cs (hunk 6 above)
Comment (grisha-kotler) [reply src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideExternalReplicationOperation.cs:10]: done

File: src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideExternalReplicationsOperation.cs (hunk 7 above)
Comment (grisha-kotler) [reply src/Raven.Client/ServerWide/Operations/Configuration/GetServerWideExternalReplicationsOperation.cs:9]: done

File: src/Raven.Client/ServerWide/Operations/Configuration/PutServerWideExternalReplicationOperation.cs (hunk 9 above)
Comment (grisha-kotler) [reply src/Raven.Client/ServerWide/Operations/Configuration/PutServerWideExternalReplicationOperation.cs:12]: done

File: src/Raven.Client/ServerWide/Operations/Configuration/ToggleServerWideTaskStateOperation.cs (hunk 10 above)
Comment (grisha-kotler) [reply src/Raven.Client/ServerWide/Operations/Configuration/ToggleServerWideTaskStateOperation.cs:9]: done

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs (hunk 12 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/ClusterStateMachine.cs:76]: done

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs (hunk 15 above)
Comment (grisha-kotler) [reply src/Raven.Server/Web/System/AdminServerWideHandler.cs:138]: done

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs (hunk 16 above)
Comment (grisha-kotler) [reply src/Raven.Server/Web/System/AdminServerWideHandler.cs:175]: It's getting all of the server-wide tasks

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs (hunk 17 above)
Comment (grisha-kotler) [reply src/Raven.Server/Web/System/AdminServerWideHandler.cs:219]: done

File: src/Raven.Server/ServerWide/Commands/PutServerWideExternalReplicationCommand.cs (hunk 20 above)
Comment (grisha-kotler) [reply src/Raven.Server/ServerWide/Commands/PutServerWideExternalReplicationCommand.cs:96]: We generate a connection string for the external replication

File: src/Raven.Client/ServerWide/Operations/Configuration/DeleteServerWideTaskOperation.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Client/ServerWide/Operations/Configuration/DeleteServerWideTaskOperation.cs:11]: done

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs (hunk 18 above)
Comment (grisha-kotler) [reply src/Raven.Server/Web/System/AdminServerWideHandler.cs:312]: done

File: src/Raven.Server/Web/System/AdminServerWideHandler.cs (hunk 19 above)
Comment (grisha-kotler) [reply src/Raven.Server/Web/System/AdminServerWideHandler.cs:331]: done

**Pull request closed without merging**